packages/config/src/validate/validations.js

```diff
@@ -149,6 +156,22 @@ const POST_NORMALIZE_VALIDATIONS = [
     ...insideRootCheck,
     example: (edgeHandlers) => ({ build: { edge_handlers: removeParentDots(edgeHandlers) } }),
   },
+  {
+    property: 'functions.*.externalModules',
```

Comment:
Should we add a "pre-normalize" validation to ensure that `functions.{key}` is an plain object, except when `key` is either `externalModules` or `ignoredModules`?
(This would also remove the need for `isConfigLeaf()` above.)

Reply:
If we do this, I think we'll end up duplicating validation logic for each key in two places: once in this pre-normalisation step, and the other for the individual keys after normalisation.

If you do specify a `functions.{key}` that is not a plain object, it'll be inferred as a configuration property of the top-level object, which is valid syntax. If the goal is to prevent unknown configuration keys (e.g. something other than `externalModules` or `ignoredModules`), wouldn't it be better to do it after normalisation, so that we can validate both the top-level and function-level objects?

I did consider this, but my understanding is that we allow unknown properties to get through elsewhere in the application?

Reply:
What about doing it in the post-normalization steps, checking that any `functions.*` is a plain object?

The downside of doing it in the post-normalization is that the normalization logic would not be able to rely on the assumption that `functions.*` is a plain object. However, it looks like the current normalization logic would work without that assumption. 

The main reason would be to validate against users using something like:

```toml
[functions]
my_function = "/functions/my_function.js"
```

Syntax like this might break `@netlify/config` because of the assumption that `functions.*` is a plain object. For example, the `functions.*.js_external_modules` validation might rely on that assumption.

Reply:
Ah, good point! I've done it in the post-normalisation step and, as you say, it works fine. I've added a test for the specific example you shared. It's in https://github.com/netlify/build/pull/2390/commits/7e33c07f033f3e34c622c453d8e280aee1199620.

Reply:
(Let me know if that looks okay to you.)

Reply:
This looks good! :+1: